src/diff/index.js

```diff
@@ -120,25 +120,47 @@ export function diff(
 				}
 
 				if (
-					!c._force &&
-					c.shouldComponentUpdate != null &&
-					c.shouldComponentUpdate(newProps, c._nextState, componentContext) ===
-						false
+					(!c._force &&
+						c.shouldComponentUpdate != null &&
+						c.shouldComponentUpdate(
+							newProps,
+							c._nextState,
+							componentContext
+						) === false) ||
+					(newVNode._original === oldVNode._original && !c._processingException)
 				) {
 					c.props = newProps;
 					c.state = c._nextState;
-					c._dirty = false;
+					/**
+					 * When the diff originates from a component we know that this was the component that dirtied the component
+					 * linearly this means that if vnodes are equal that an update could've originated from this component and is still
+					 * queued for an update.
+					 * Example:
+					 *
+					 * We have a wrapping component, when this wrapping component updates its internal state the children
+					 * ref could have remained static:
+					 *
+					 * <div>
+					 *   {children}
+					 * </div>
+					 *
+					 * What if inside this children ref there was another setState then this should not be marked as not dirty yet until
+					 * we have traversed this node.
+					 */
+					if (newVNode._original !== oldVNode._original) c._dirty = false;
```

Comment:
Imagine in the context of https://github.com/preactjs/preact/pull/2386/files#diff-54682ce380935a717e41b8bfc54737f6R68 that the `Provider` updates internal state but `Layout` does so as well, this would imply that we `diff` from `Provider` end up at `Layout` see the nodes are equal and mark it as clean while it has pending state that could lead to erroneous DOM